# <span style = "color:rebeccapurple">Object Oriented Programming</span>
**(Or how to be Gordon Ramsay in under an hour)**

Yesterday you learned how to write your own functions. This is one of the most important skills since functions are the basis of programmatic action. Today we will learn how to write our own `classes`. Classes are usually a type of specialized data or a combination, within one object, of data and a bundle of actions.

Classes are the basis for what is known as Object Oriented Programming (which I like to call OOPs), where you create entities called **objects**, each of which has a different set of skills and data. You then ask these objects to perform what you want to be performed according to their skills. Since research projects have a high degree of complexity, mutability, recycling and remixing, OOP comes in handy to organize data and action better.

## <span style = "color:darkorchid">Concepts of Object Oriented Programming</span>

The backbone of OOP are, of course, **objects**. An object is an *instance* of a **class**. For example a specific dorito you eat is an instance of a triangle. The definition of a triangle is the class. All doritos are instances of the class triangle. Think of classes as the abstract blueprints of objects.

To better understand objects and classes, we'll have to visit Gordon Ramsay.

Imagine you are at a high-end restaurant. Let's say you are seeing Gordon Ramsay at work. There is an executive chef, a head chef, several sous-chefs, specialized chefs (for roasting, for pastries, etc.), each with a team of specialists (the butcher, the grill chef, the baker, the confectioner, etc.). Here is an image I got from google images:

<img src="images/brigade-de-cuisine-high-speed-learning.png" width="35%">
<!-- ![brigade-de-cuisine-high-speed-learning.png](images/brigade-de-cuisine-high-speed-learning.png) -->

Think of each of these as a type or a *class* of chefs. There are important things to note:
<ul>
    <li>They are all of the generic type <i>chef</i>, but some have extra skills or responsibilities.</li>
    <li>You can have several chefs of the same class (like several sous-chefs). However, these are not the same people!</li>
</ul>

Well, classes in python are something similar. It is a specified type of entity that has specific skills and attributes. Objects are the realizations of these classes. Each realization is called an **<span style = "color:blue">instance</span>**. In python, the skills are called **<span style = "color:blue">methods</span>**, these are actions that all instances of a class can perform. They also have **<span style = "color:blue">attributes</span>** which are variables, possibly unique, that each instance has (like the names of individual chefs).

### <span style = "color:teal">The Gordon Ramsay Principle</span>

When you are writing a large python project, do not think of yourself as a homecook that does everything by themselves from scratch, following each step one after another. Instead, **<span style = "color:crimson">think of yourself as an executive chef</span>**. First you appoint all the chefs working for you, each with predetermined skills and attributes. You think of the overall plan, and then you delegate tasks to the respective chefs (which may, in turn, delegate tasks to their respective chefs and specialists).

## <span style = "color:darkorchid">OOP In Action</span>

Before creating our own classes, let's revisit some old friends.

### <span style = "color:teal">Predefined classes</span>

You are already familiar with built-in Python object classes like *strings*, *integers*, *floats*, *lists*, and *dictionaries*. Let's see them in a new light.

Take strings as an example. In the code below, `my_string` is an object **instance** of the **class** `string`.

In [1]:
my_string = "Hello World."
print(my_string)

Hello World.


And in the code below, `another_string` is also an object **instance** of **class** `string`.

In [2]:
another_string = "Hello again."
print(another_string)

Hello again.


`my_string` and `another_string` are different objects, but they are both strings. That means they have the same methods (they can perform the same actions), and they have to follow the same rules all strings must follow.

We've already encountered some string methods:

In [3]:
my_string.lower()

'hello world.'

In [4]:
my_string.upper()

'HELLO WORLD.'

In [5]:
my_string.capitalize()

'Hello world.'

In [6]:
"_".join(["hello", "world"])

'hello_world'

Note that `.join()` is a method of a string, not of lists. The one performing the "joining" is the underscore string `_`. The list is an input to the method.

What if we tried to `.lower()` a non-string object?

In [7]:
some_number = 2
some_number.lower()

AttributeError: 'int' object has no attribute 'lower'

So if a class does not have a given method, you will get an error. That object doesn't know how to perform what you are asking of it!

### <span style = "color:teal">Methods and Attributes</span>

We mentioned methods are the *actions* an object can perform, while *attributes*  are its properties.

Above, what was the recipe for `.lower()`, `.join()`, etc? Well, they all end with a parenthesis `()`, and they follow the object with a period `.`. The syntax is this:

``object.method()``.

Sometimes the methods have inputs, like in the case of `.join(some_list)`, which receives a list to join.

How can you remember this? Well remember when you call a function you use the parenthesis. Since methods are like functions for objects (the action element), this should help.

What happens if you forget the parenthesis?

In [8]:
my_string.lower

<function str.lower()>

What's that? It actually returned a full function, as opposed to its output. In order for the function to "activate" and perform what it needs to, you must add the parenthesis.

Now what about attributes? `str` objects actually don't have any attributes, they are too simple! Indeed, most objects we've dealt with (numbers, lists, etc.) don't have attributes (besides their own values). Some more complex objects (like files and modules) do, but we won't go into those. We will instead create our own classes with attributes!

## <span style = "color:darkorchid">Defining Our Own Classes</span>

To define your own class, you just need the keyword `class`, use the following syntax:

In [9]:
class MySillyClass():
    pass

The `pass` statement is a placeholder that allows us to not do anything. Our class exists, but it doesn't have any attribute or methods.

To create an instance object of the class `MySillyClass` we do the following:

In [10]:
silly_object = MySillyClass()

Cool. `silly_object` can't do anything, and it doesn't have any data, but we can still see its class:

In [11]:
silly_object

**Note**

- The convention is to use **CamelCase** for class names. Note this is different to the **snake_case** we used for variable and function names.
- When defining a class with the `class MyClass():` syntax, you don't actually need the parenthesis. There are cases where we will need them (later today), but it is OK if you don't use them.
- To create a new object of a given class, however, you need the parentheses. Type the class name and a parenthesis. Don't forget the parenthesis!!

OK, let's get back to our culinary dreams. Let us build the pastry chef. We'll start with the class definition:

In [12]:
class PastryChef:
    pass

It doesn't do anything.

### <span style = "color:teal">Methods</span>

To add a method, we use the function syntax **inside** the class definition.

Let's redefine the `PastryChef` class and give it the method `bake`:

In [13]:
class PastryChef:
    def bake(self, food):
        print(f"I've baked your {food}!")

We'll inspect these elements in one second, but first let's see what this does:

In [14]:
# Create a pastry chef, 
cerise = PastryChef()

In [15]:
cerise.bake("cookie")

I've baked your cookie!


<hr style = "color:red"/>

**<span style = "color:red">EXERCISE 1 - Astronomy Application</span>**

Let's revisit what we learned yesterday when creating our own functions.

Create a class called `PlanetEscape` with one method called `calculate_escape_velocity`. It should receive two inputs: `mass` (in Earth masses) and `radius` (in Earth radii). It should return the escape velocity in km/s using the formula from yesterday's quiz:

$$v_{esc} = \sqrt{\frac{2GM}{R}}$$

Where $G = 1.541 \times 10^{-6}\text{ R}_\oplus^3 \text{ M}_\oplus^{-1} \text{ s}^{-2}$. You will need to multiply the result by Earth's radius (6,371 km) to convert to km/s.

It's OK to go back to yesterday's notes for help.

In [16]:
class PlanetEscape():
    def calculate_escape_velocity(self, planet_mass, planet_radius):
        G = 1.541e-6
        v_esc = ((2 * G * planet_mass) / planet_radius)**(1/2)
        v_esc_kms = 6371 * v_esc
        return v_esc_kms

In [17]:
escape = PlanetEscape()
escape.calculate_escape_velocity(1,1)

11.184689247448942

<hr style = "color:red"/>

**``self``**

Did you notice when we defined our bake method I had a first argument called `self`? This is done so the obejct can refer to itself and its internal methods and attributes (which we will see in one second). We will see more uses of it in a second and you will get familiar with it.
- You actually don't have to use the word "self", you could use any other word, but everybody uses and understands  "self", so please use it.
- When calling a method outside of the class definition, like in `cerise.bake("cookie")`, we completely ignore the "self" argument.

**``__init__``**

There is an important method that always exists in the background, even if you don't define it. This is `__init__`. What this does is it gives python specifications of how to initialize the object. When we define our own classes, the first method we should define is `__init__`. So let's review our `PastryChef` class:

In [18]:
class PastryChef:
    def __init__(self):                             # <-- Note we still use self here!
        print("You have created a Pastry Chef!!")

In [19]:
cerise = PastryChef()

You have created a Pastry Chef!!


We can have more than one method. In this case, we'll have both ``__init__()`` and `bake()`.

In [20]:
class PastryChef:
    def __init__(self):                             # <-- Note we still use self here!
        print("You have created a Pastry Chef!!")
        
    def bake(self, food):
        print(f"I've baked your {food}!")

### <span style = "color:teal">Attributes</span>

Let's give our chef some attributes. This are variables containing data. Let's start with something simple, a name. This is where `self` comes in handy:

In [21]:
class PastryChef:
    def __init__(self, given_name):                         # <-- A new argument
        self.name = given_name
        print("You have created a Pastry Chef!!")
        
    def bake(self, food):
        print(f"I've baked your {food}!")

In [22]:
cerise = PastryChef("Cerise Crumble")

You have created a Pastry Chef!!


**Note**

Now, `__init__` is expecting an argument when the object is initialized. If we try to initialize it without a name. What would it do?

In [23]:
PastryChef()

TypeError: PastryChef.__init__() missing 1 required positional argument: 'given_name'

**Accessing Attributes**.

To access an attribute of an object, you use the period `.` followed by the attribute name, but without parenthesis. This resembles variable names, as opposed to functions:

In [24]:
cerise.name

'Cerise Crumble'

In [25]:
cerise.bake("cookie")

I've baked your cookie!


**Modifying Attributes**

You can modify an attribute as you would for a normal variable:

In [26]:
cerise.name = "Cerise Ganache"

In [27]:
cerise.name

'Cerise Ganache'

**`dir( )`**

Before continuing, you will often need to know what methods and attributes an object has. You can use the built-in function `dir()` to find out.

In [28]:
dir(cerise)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'bake',
 'name']

Well that was unexpected.

We only defined `__init__`, `bake()`, and `name`, so what's going on? All the attributes surrounded by double underscores are predefined and all classes have them. For now we have only modified `__init__`, and is the only one we will learn about today.

We won't worry about the rest, they are not particularly obscure, we just don't have time. As you get more familiar with classes you will learn about these on your own.

In [29]:
class PastryChef:
    def __init__(self, given_name):                         # <-- A new argument
        self.name = given_name
        print("You have created a Pastry Chef!!")
        
    def bake(self, food):
        print(f"I've baked your {food}!")

    def introduction(self):
        message = f"My name is {self.name}"
        print(message)

In [30]:
vg = PastryChef("Van Gogh")
vg.introduction()

You have created a Pastry Chef!!
My name is Van Gogh


<hr style = "color:red"/>

**<span style = "color:red">EXERCISE 2</span>**

- Create a `PlanetData` class. It should take three inputs in its `__init__` method: its `name`, its `mass` (in Earth masses), and its `radius` (in Earth radii). These should be assigned to its "self" attributes `name`, `mass`, and `radius`.

- Check that you can initialize planets (like Earth with mass=1, radius=1) and access their attributes.

In [ ]:
class PlanetData():
    def __init__(self, name, mass, radius):
        self.name = name
        self.mass = mass
        self.radius = radius
        

In [32]:
earth_data = PlanetData("earth", 1,1,)

In [33]:
earth_data.name

'earth'

<hr style = "color:red"/>

### <span style = "color:teal">Class Attributes vs Instance Attributes</span>

The attribute `name` we defined above is an instance attribute (remember we assigned it to the instance using `self`). As it turns out, there are attributes that the whole class shares. This is how you can define those:

In [34]:
class PastryChef:
    specialization = "Pastry"
    
    def __init__(self, given_name):
        self.name = given_name
        # I'll stop printing initialization method from now on.
        
    def bake(self, food):
        print(f"I've baked your {food}!")

In [35]:
cerise = PastryChef("Cerise Crumble")
claire = PastryChef("E. Claire Framboise")

In [36]:
print(cerise.name, ", Specialization: ", cerise.specialization)

print(claire.name, ", Specialization: ", claire.specialization)

Cerise Crumble , Specialization:  Pastry
E. Claire Framboise , Specialization:  Pastry


You have to be careful with class attributes, because these can potentially be shared, like this:

In [37]:
class PastryChef:
    specialization = []
    def __init__(self, given_name):
        self.name = given_name

In [38]:
cerise = PastryChef("Cerise")

claire = PastryChef("Claire")

In [39]:
# Print specialization
print(cerise.specialization)
print(claire.specialization)

# Modify individual specialization:
cerise.specialization.append("Confectionery")
claire.specialization.append("Bakery")

# Print specialization again
print(cerise.specialization)
print(claire.specialization)

[]
[]
['Confectionery', 'Bakery']
['Confectionery', 'Bakery']


Woah! Cerise and Claire now share the same list of specializations, which was not our intention.

### <span style = "color:teal">More on methods</span>

Do you remember the `self` argument to our methods? Well, it will help us access attributes inside methods. Say we want an "introduction" method, where the chefs introduce themselves. They need to know their own name!

In [ ]:
class PastryChef:
    def __init__(self, given_name):
        self.name = given_name
    
    def introduction(self):
        print(f"Hello, my name is {self.name}.")

In [ ]:
chuck = PastryChef("Chuck O'Leith")

In [ ]:
chuck.introduction()

Note the `introduction()` method, inside the class definition, only has the self argument, which is how we access the `name` attribute. When calling this method for `chuck`, no argument is needed.

OK, let's start building more interesting methods. I want to calculate the volume of cakes, so I'll need the module `math`.

In [40]:
import math

In [41]:
class PastryChef:

    def __init__(self, given_name):
        self.name = given_name
    
    def get_cake_volume(self, n_people):
        height = 3
        radius = math.sqrt(n_people)
        volume = height * (math.pi * radius**2)
        return volume

In [42]:
mack = PastryChef("Mack A. Roon")
mack.get_cake_volume(n_people = 4)

37.69911184307752

Now let's add a method that recommends a dessert based on your desired quality:

In [43]:
class PastryChef:
    menu_recommendations = {
        'fun': 'banana split',
        'fried': 'churros',
        'flaky': 'croissant',
        'frozen': 'ice cream',
        'fudgy': 'brownie',
        'flavorful': 'tiramisu',
        'finessed': 'opera cake'
    }

    def __init__(self, given_name):
        self.name = given_name
    
    def get_cake_volume(self, n_people):
        height = 3
        radius = math.sqrt(n_people)
        volume = height * (math.pi * radius**2)
        return volume
    
    def recommend(self, quality):
        print(f"Based on your desired quality, I recommend {self.menu_recommendations[quality]}")

In [46]:
mack = PastryChef("Mack A. Roon")
mack.recommend("flavorful")

Based on your desired quality, I recommend tiramisu


What if we ask for a non-f quality?

In [ ]:
class PastryChef:
    menu_recommendations = {
        'fun': 'banana split',
        'fried': 'churros',
        'flaky': 'croissant',
        'frozen': 'ice cream',
        'fudgy': 'brownie',
        'flavorful': 'tiramisu',
        'finessed': 'opera cake'
    }

    def __init__(self, given_name):
        self.name = given_name
    
    def get_cake_volume(self, n_people):
        height = 3
        radius = math.sqrt(n_people)
        volume = height * (math.pi * radius**2)
        return volume
    
    def recommend(self, quality):
        try:
            print(f"Based on your desired quality, I recommend {self.menu_recommendations[quality]}")
        except KeyError:
            print("Fanciful Folks Favored. Forego Further Forays.")

In [48]:
mack = PastryChef("Mack A. Roon")
mack.recommend("finessed")
mack.recommend("tart")

Based on your desired quality, I recommend opera cake
Fanciful Folks Favored. Forego Further Forays.


In [49]:
mack.recommend([])

TypeError: unhashable type: 'list'

<hr style = "color:red"/>

**<span style = "color:red">EXERCISE 3</span>**

Fuse `PlanetEscape` and `PlanetData` into one `Planet` class:
- Modify the `calculate_escape_velocity` method so that it no longer accepts `mass` and `radius` as arguments. Instead, it should read these values directly from the planet's own attributes (`self.mass` and `self.radius`)!

- Then, add a new method called `report_data` that prints a statement summarizing the planet's properties. For example, calling `.report_data()` on an Earth object should print: "Planet Earth has a mass of 1 Earth masses and a radius of 1 Earth radii. Its escape velocity is 11.19 km/s."

In [50]:
class Planet():
    G = 1.541e-6
    def __init__(self, name, mass, radius):
        self.name = name
        self.mass = mass
        self.radius = radius

    def calculate_escape_velocity(self):
        v_esc = ((2 * self.G * self.mass) / self.radius)**(1/2)
        v_esc_kms = 6371 * v_esc
        return v_esc_kms
    
    def report_data(self):
        v = self.calculate_escape_velocity()
        print(f"name is {self.name} and v_esc is {v}")

In [51]:
earth = Planet("earth", 1, 1)
earth.report_data()

name is earth and v_esc is 11.184689247448942


<hr style = "color:red"/>

Now let's bring back our `.bake()` method, and add a `prepare_cake()` method:

In [ ]:
class PastryChef:
    menu_recommendations = {
        'fun': 'banana split',
        'fried': 'churros',
        'flaky': 'croissant',
        'frozen': 'ice cream',
        'fudgy': 'brownie',
        'flavorful': 'tiramisu',
        'finessed': 'opera cake'
    }

    def __init__(self, given_name):
        self.name = given_name
    
    def get_cake_volume(self, n_people):
        height = 3
        radius = math.sqrt(n_people)
        volume = height * (math.pi * radius**2)
        return volume
    
    def recommend(self, quality):
        try:
            print(f"Based on your desired quality, I recommend {self.menu_recommendations[quality]}")
        except KeyError:
            print("Fanciful Folks Favored. Forego Further Forays.")
    
    def bake(self, cake, volume):
        print(f"I've baked your {cake} cake, it was {volume} units.")
    
    def prepare_cake(self, n_people, cake):
        volume = self.get_cake_volume(n_people)
        self.bake(cake, volume)

In [ ]:
cerise = PastryChef("Cerise Crumble")

In [ ]:
# Ask Cerise for a recommendation
cerise.recommend("fried")

In [ ]:
# Ask Cerise to make a cake for 5 people
cerise.prepare_cake(n_people=5, cake="velvet")

## <span style = "color:darkorchid">Review</span>

The recipe for making your own classes can be summarized like this:

In [ ]:
class MyClass:
    class_attribute = "something here, could be string, number, etc."

    def __init__(self, att1, att2):
        self.attribute1 = att1
        self.attribute2 = att2
    
    def do_something(self, arg1):
        # This method does something
        # ...
        return
    
    def do_something_else(self):
        # you can call your own methods:
        self.do_something(self.attribute1)
        # and later do something else ...
        # ...
        return

<hr style = "color:red"/>

**<span style = "color:red">EXERCISE 4</span>**

Let's go back to your `Planet` class.
- Add a class attribute `G` set to `1.541e-6`.
- Update `__init__()` to initialize an instance attribute `moons` as an empty list.
- Add a method `add_moon` that takes a moon name (string) as an input and appends it to the planet's list of moons.
- Add a method `list_moons` that prints a message like "Moons of {self.name}: {list of moons}". If the planet has no moons, print a message saying so.

Test your updated class by creating Jupiter (mass=317.8, radius=11.2), adding a few of its moons (Io, Europa, Ganymede, Callisto), and listing them. Remember to include `return` at the end of all methods!

<hr style = "color:red"/>

## <span style = "color:darkorchid">Case Examples</span>

### <span style = "color:teal">Astronomy</span>

**Celestial Body Representation**:

You can represent stellar and planetary bodies as objects. A `Star` or `Planet` object contains physical attributes like mass, radius and spectral type, with methods to compute derived values like luminosity, escape velocity, or orbital period. This keeps related data and physics together.

**Astropy Coordinate Conversions**:

Widely-used astronomy python libraries like `astropy` and rely heavily on OOP. For instance, `astropy.coordinates.SkyCoord` is a class representing sky coordinates. You create instances of `SkyCoord` and use methods like `.transform_to('galactic')` to convert coordinate systems effortlessly.

It is worth noring that astropy has great unit conversion utilities and all the constants, so if you find the units confusing (and many people do), make sure to investigate astropy!!

**N-body Simulations**:

When simulating the gravitational dynamics of star clusters or solar systems, each body is modeled as an instance of a `Particle` class, which tracks its own 3D position, velocity, and mass, and calculates forces during integration steps.

### <span style = "color:teal">A Machine Learning Project</span>

A common pipeline in machine learning includes the following elements and/or stages: data collection, pre-processing, data splitting, optimization, model selection, testing, post-processing, visualization.

* Your data types may be very different from each other.
* These steps do not always happen in this order.
* We may have different optimizers that later ensemble in some way.
* We may have multiple models interacting with each other.
* etc.

Therefore, it is beneficial to define classes that perform each of these stages (or sub-stages within), and create instances of those classes as we need them. For example, a class for each data type (a time series, real-valued samples, networks, data with symmetric properties, etc.), a class for different types of pre-processors, a class for each optimizer, etc.

Popular machine learning libraries like transformers, pytorch, tensorflow, scikit-learn, etc. have pre-determined classes that specialize in particulars tasks.

### <span style = "color:teal">A Simulation Project</span>
Simulation paradigms benefit greatly from OOP. For example, in agent based modeling, you simulate interactions among individual agents. In this case, each agent must exist by itself, and it has a variety of internal properties and abilities.

This principle extends widely: used in economics, biology, epidemiology, astronomy, transportation, etc.
Newer machine learning approaches leverage physical principles while learning. In this case two different "paradigms" collide: standard machine learning and (often) differential equations. You will save yourself much with specialized classes that deal with each component.

### <span style = "color:teal">Genomics</span>

The well-known biopython package is object oriented. Each gene sequence belongs to the class of gene sequences. If they are transcribed to protein sequences, they are their own protein sequence class! Etc.

<hr style = "color:red"/>

### <span style = "color:red">EXERCISE</span>

Think about a common pipeline in your own field. How would you conceptualize this using OOP? Which classes would you have? Which methods and attributes would these classes have? How would they interact?

<blockquote>
Write your thoughts here.

<hr style = "color:red"/>

## <span style = "color:red">Final Quiz</span>

This quiz puts what we learned yesterday and today in practice. It's a bit long, so we will use an LLM's help to get the full answer. But I want you to take 5 minutes to write **in paper** a plan of how you would accomplish it (what classes do we need? what class attributes and what instance attributes? what methods? etc.). I also recommend drawing a diagram of how classes or instances connect to each other, what each contains, etc.

### <span style = "color:teal">Tasks</span>

Write a class called `KeplerSystem` that models a stellar system and applies Kepler's laws using classes. Your `KeplerSystem` object will contain and interact with instances of your `Planet` class!

1. **Star Properties**: The `KeplerSystem` should be initialized with the `star_name` and the `star_mass` (in Earth masses).

2. **Orbiters Database**: Keep a list of planets orbiting this star, initialized as an empty list (e.g., `self.orbiting_planets = []`).

3. **Kepler's Third Law**: For each planet it contains, your `KeplerSystem` should be able to calculate the semi-major axis $a$ in astronomical units (AU) using Kepler's Third Law:
   $$a = \left(\frac{G(M_{star} + m_{planet}) T^2}{4\pi^2}\right)^{1/3}$$

- Both $M_{star}$ and $m_{planet}$ are in Earth masses, so no unit conversion
     is needed. Use $G = 8.888 \times 10^{-10}\text{ AU}^3\text{ M}_\oplus^{-1}\text{ day}^{-2}$.
     
     (Note: This differs from the previous day where we were using Solar masses. We're using Earth masses today to stay consistent with exercises 1-4)

4. **Adding Orbiters**: Add a method `add_planet(self, planet, period_days)` that takes two inputs:
   - `planet`: An instance of your `Planet` class.
   - `period_days`: The orbital period of that planet in Earth days.
   - Calculate the semi-major axis, and adds it as a new attribute on the planet instance (e.g., `planet.semi_major_axis_au = sma`).
   - Appends the planet instance to the `orbiting_planets` list.

5. **System Summary**: Add a method `get_system_summary(self)` that prints a formatted summary of the stellar system. It should show the star's name and mass, and list each orbiting planet, its mass, its escape velocity, and its calculated orbital semi-major axis in AU.

Test your class by creating a system for our Sun (`star_mass = 333000.0`, since
the Sun is approximately 333,000 Earth masses), creating a Planet instance for
Earth (`mass = 1.0`, `radius = 1.0`), adding it to the system with a period of
`365.25` days, and calling `get_system_summary()`. The calculated semi-major axis
should be approximately `1.00` AU!

In [52]:
import math

class Planet:
    G = 1.541e-6
    
    def __init__(self, name, mass, radius):
        self.name = name
        self.mass = mass
        self.radius = radius
        # This will be set dynamically by KeplerSystem
        self.semi_major_axis_au = None

    def calculate_escape_velocity(self):
        v_esc = ((2 * self.G * self.mass) / self.radius)**(1/2)
        v_esc_kms = 6371 * v_esc
        return v_esc_kms
    
    def report_data(self):
        v = self.calculate_escape_velocity()
        print(f"name is {self.name} and v_esc is {v}")


class KeplerSystem:
    # Constant G provided for the system calculation
    G_CONSTANT = 8.888e-10

    def __init__(self, star_name, star_mass):
        self.star_name = star_name
        self.star_mass = star_mass  # In Earth masses
        self.orbiting_planets = []  # List to store Planet instances

    def add_planet(self, planet, period_days):
        """
        Calculates the semi-major axis using Kepler's Third Law, 
        assigns it to the planet, and tracks the planet in the system.
        """
        # Kepler's Third Law calculation
        numerator = self.G_CONSTANT * (self.star_mass + planet.mass) * (period_days ** 2)
        denominator = 4 * (math.pi ** 2)
        semi_major_axis = (numerator / denominator) ** (1 / 3)
        
        # Dynamically add the attribute to the planet instance
        planet.semi_major_axis_au = semi_major_axis
        
        # Append the planet to the orbiters database
        self.orbiting_planets.append(planet)

    def get_system_summary(self):
        """Prints a formatted summary of the stellar system."""
        print("=" * 50)
        print(f"STELLAR SYSTEM SUMMARY: {self.star_name.upper()}")
        print("=" * 50)
        print(f"Star Name: {self.star_name}")
        print(f"Star Mass: {self.star_mass:,} Earth masses")
        print(f"Number of Orbiting Planets: {len(self.orbiting_planets)}")
        print("-" * 50)
        
        for idx, planet in enumerate(self.orbiting_planets, 1):
            v_esc = planet.calculate_escape_velocity()
            print(f"  {idx}. Planet: {planet.name}")
            print(f"     Mass: {planet.mass} Earth masses")
            print(f"     Escape Velocity: {v_esc:.2f} km/s")
            print(f"     Semi-Major Axis: {planet.semi_major_axis_au:.2f} AU")
            print("-" * 50)


# ==========================================
# Testing the Implementation
# ==========================================
if __name__ == "__main__":
    # 1. Create the stellar system for our Sun
    sol_system = KeplerSystem(star_name="Sun", star_mass=333000.0)

    # 2. Create the Planet instance for Earth
    earth = Planet(name="Earth", mass=1.0, radius=1.0)

    # 3. Add Earth to the system with an orbital period of 365.25 days
    sol_system.add_planet(earth, period_days=365.25)

    # 4. Generate and print the system summary
    sol_system.get_system_summary()

STELLAR SYSTEM SUMMARY: SUN
Star Name: Sun
Star Mass: 333,000.0 Earth masses
Number of Orbiting Planets: 1
--------------------------------------------------
  1. Planet: Earth
     Mass: 1.0 Earth masses
     Escape Velocity: 11.18 km/s
     Semi-Major Axis: 1.00 AU
--------------------------------------------------


<hr style = "color:red"/>

## <span style = "color:darkorchid">(Take-home material) Inheritance</span>

Assume that all pastry chefs can do what confectioners and bakers can, but not the other way around. How do we deal with this situation? Well, we could define three different classes, each with their own methods. However, we would be repeating a lot of code in the definition of pastry chefs. Also, we wouldn't know, unless we check our code line by line, that pastry chefs can what bakers and confectioners can.

Enter the notion of **inheritance**. We can have a base class with given methods and attributes, and another class that *inherits* those methods and attributes. This second class we call the *child* class of the first (which we call the *parent* class).

Implementing is not difficult. When we define the class, use the parenthesis in the class name and write the parent class. Let's see this in action.

In [ ]:
class Baker():
    def __init__(self, name):
        self.name = name
    
    def bake(self, cake):
        print(f"Your {cake} cake is ready!!")

In [ ]:
claire = Baker("Claire")
claire.bake("velvet")

In [ ]:
class PastryChef(Baker):        # <-- Note the Baker indication
    pass

In [ ]:
cerise = PastryChef("Cerise")

In [ ]:
cerise.name

In [ ]:
cerise.bake("velvet")

Woah! What happened here? Notice we did not do anything in the `PastryChef` definition, and Cerise still had a name and knew how to bake. That's because it inherited those skills from its base / parent class.

What if we want to modify the `__init__()` method?

In [ ]:
class PastryChef(Baker):
    def __init__(self, name):
        super().__init__(name)
        print("Pastries Rock!")

OK what happened here? Why are there two ``__init__()``s? When we define the child class initialization, we must tell it to first intialize what has to be initialized in the parent class. Then it can perform other things, in this case, print a message.

In [ ]:
cerise = PastryChef("Cerise")

We can also add methods specific to pastry chefs:

In [ ]:
class PastryChef(Baker):
    def __init__(self, name):
        super().__init__(name)
    
    def create_recipe(self):
        print("Your recipe is...")

Note that while Pastry chefs can create a recipe, bakers can't:

In [ ]:
cerise = PastryChef("Cerise")
cerise.create_recipe()

In [ ]:
claire.create_recipe()

We can also have head chefs that can do everything a Pastry chef can (this is not true in real life, it's just an example):

In [ ]:
class HeadChef(PastryChef):
    pass

In [ ]:
cerise = HeadChef("Cerise")
print(type(cerise))
cerise.bake("velvet")
cerise.create_recipe()

Notice how in terms of class inheritance a HeadChef class is the "granddaughter" of a Baker class. This seems "backward" with respect to our hierarchy tree image. So pay special attention when designing inheritance classes.

OK, now what if we have a confectioner, and pastry chefs can also do what those can? Enter **Multiple Inheritance**.

In [ ]:
class Confectioner():
    def confection(self):
        print("I have confectioned ... I think?")

In [ ]:
class PastryChef(Baker, Confectioner):
    pass

In [ ]:
claire = Baker("Claire")
chuck = Confectioner()
cerise = PastryChef("Cerise")

In [ ]:
claire.bake("velvet")

In [ ]:
claire.confection()
# this should raise an error, why?

In [ ]:
chuck.confection()

In [ ]:
chuck.bake("velvet")
# this should raise an error, why?

In [ ]:
cerise.bake("velvet")

In [ ]:
cerise.confection()

Inheritance and multiple inheritance are extremely useful, so while we can't continue further, I encourage you to learn more and practice them in your own projects!